File: devtools/bench/warehouse/warehouse.swift
// Warehouse robot simulation — the Swift half of the Saw-vs-Swift comparison.
// Idiomatic Swift OO: final classes for the coordinating objects (robots hold
// Order references through their mission enum, ARC manages them), a struct
// RNG, Array storage. The simulation logic, traversal order, and tie-breaks
// mirror warehouse.saw exactly; the printed checksums must match it.

let W = 64
let ROBOTS = 100
let TICKS = 200_000
let ORDER_EVERY = 1
let PENDING_CAP = 500
let BATTERY_CAP = 1200
let BATTERY_LOW = 150
let CHARGE_RATE = 25

struct Rng {
    var state: UInt64
    mutating func next() -> UInt64 {
        state = state &* 6364136223846793005 &+ 1442695040888963407
        return state >> 33
    }
    mutating func below(_ n: Int) -> Int {
        Int(next() % UInt64(n))
    }
}

final class Order {
    let id: Int
    let px: Int
    let py: Int
    let dx: Int
    let dy: Int
    init(id: Int, px: Int, py: Int, dx: Int, dy: Int) {
        self.id = id
        self.px = px
        self.py = py
        self.dx = dx
        self.dy = dy
    }
}

enum Mission {
    case idle
    case toPickup(Order)
    case toDropoff(Order)
    case charging
}

final class Robot {
    var x: Int
    var y: Int
    var battery: Int
    var delivered = 0
    var mission = Mission.idle
    init(x: Int, y: Int, battery: Int) {
        self.x = x
        self.y = y
        self.battery = battery
    }
    var isIdle: Bool {
        if case .idle = mission { return true }
        return false
    }
}

func dist(_ ax: Int, _ ay: Int, _ bx: Int, _ by: Int) -> Int {
    (ax > bx ? ax - bx : bx - ax) + (ay > by ? ay - by : by - ay)
}

final class Dispatcher {
    var robots: [Robot] = []
    var created = 0
    var pending: [Order] = []
    var pendingHead = 0
    var rng: Rng
    var delivered = 0
    var steps = 0

    init(rng: Rng) {
        self.rng = rng
    }

    // Moves robot one step toward (tx, ty); reports whether it now stands there.
    func moveRobot(_ r: Robot, _ tx: Int, _ ty: Int) -> Bool {
        if r.x == tx && r.y == ty { return true }
        if r.x != tx {
            if r.x < tx { r.x += 1 } else { r.x -= 1 }
        } else {
            if r.y < ty { r.y += 1 } else { r.y -= 1 }
        }
        r.battery -= 1
        steps += 1
        return r.x == tx && r.y == ty
    }

    func assignPending() {
        while pendingHead < pending.count {
            let order = pending[pendingHead]
            var best = -1
            var bestd = 1_000_000
            for r in 0..<ROBOTS {
                if robots[r].isIdle {
                    let d = dist(robots[r].x, robots[r].y, order.px, order.py)
                    if d < bestd {
                        bestd = d
                        best = r
                    }
                }
            }
            if best < 0 { return }
            robots[best].mission = .toPickup(order)
            pendingHead += 1
        }
    }

    func stepRobot(_ i: Int) {
        let r = robots[i]
        switch r.mission {
        case .idle:
            if r.battery < BATTERY_LOW {
                r.mission = .charging
            }
        case .toPickup(let o):
            if moveRobot(r, o.px, o.py) {
                r.mission = .toDropoff(o)
            }
        case .toDropoff(let o):
            if moveRobot(r, o.dx, o.dy) {
                r.mission = .idle
                r.delivered += 1
                delivered += 1
            }
        case .charging:
            let sx = r.x < 32 ? 0 : 63
            let sy = r.y < 32 ? 0 : 63
            if moveRobot(r, sx, sy) {
                r.battery += CHARGE_RATE
                if r.battery >= BATTERY_CAP {
                    r.battery = BATTERY_CAP
                    r.mission = .idle
                }
            }
        }
    }

    func tick(_ t: Int) {
        if t % ORDER_EVERY == 0 && pending.count - pendingHead < PENDING_CAP {
            let px = rng.below(W)
            let py = rng.below(W)
            let dx = rng.below(W)
            let dy = rng.below(W)
            pending.append(Order(id: created, px: px, py: py, dx: dx, dy: dy))
            created += 1
        }
        assignPending()
        for i in 0..<ROBOTS {
            stepRobot(i)
        }
    }
}

var rng = Rng(state: 42)
var robots: [Robot] = []
for _ in 0..<ROBOTS {
    let x = rng.below(W)
    let y = rng.below(W)
    robots.append(Robot(x: x, y: y, battery: BATTERY_CAP))
}
let sim = Dispatcher(rng: rng)
sim.robots = robots

for t in 0..<TICKS {
    sim.tick(t)
}

var posHash = 0
var batterySum = 0
for i in 0..<ROBOTS {
    posHash = posHash &* 31 &+ (sim.robots[i].x * 64 + sim.robots[i].y)
    batterySum += sim.robots[i].battery
}
print("created \(sim.created)")
print("delivered \(sim.delivered)")
print("steps \(sim.steps)")
print("pos \(posHash)")
print("battery \(batterySum)")

File: devtools/bench/warehouse/warehouse.rs
// Warehouse robot simulation — the Rust half of the comparison.
// Idiomatic Rust: plain structs in a Vec, a Copy mission enum carrying an
// order INDEX (the natural Rust design — no Rc<Order> needed), field-disjoint
// &mut self methods. Logic, traversal order, and tie-breaks mirror
// warehouse.saw exactly; the printed checksums must match it.

const W: i64 = 64;
const ROBOTS: usize = 100;
const TICKS: i64 = 200_000;
const ORDER_EVERY: i64 = 1;
const PENDING_CAP: usize = 500;
const BATTERY_CAP: i64 = 1200;
const BATTERY_LOW: i64 = 150;
const CHARGE_RATE: i64 = 25;

struct Rng {
    state: u64,
}

impl Rng {
    fn next(&mut self) -> u64 {
        self.state = self
            .state
            .wrapping_mul(6364136223846793005)
            .wrapping_add(1442695040888963407);
        self.state >> 33
    }
    fn below(&mut self, n: i64) -> i64 {
        (self.next() % (n as u64)) as i64
    }
}

#[derive(Clone, Copy)]
enum Mission {
    Idle,
    ToPickup(usize),
    ToDropoff(usize),
    Charging,
}

struct Order {
    px: i64,
    py: i64,
    dx: i64,
    dy: i64,
}

struct Robot {
    x: i64,
    y: i64,
    battery: i64,
    delivered: i64,
    mission: Mission,
}

fn dist(ax: i64, ay: i64, bx: i64, by: i64) -> i64 {
    (ax - bx).abs() + (ay - by).abs()
}

struct Sim {
    robots: Vec<Robot>,
    orders: Vec<Order>,
    pending: Vec<usize>,
    pending_head: usize,
    rng: Rng,
    delivered: i64,
    steps: i64,
}

impl Sim {
    fn robot_idle(&self, r: usize) -> bool {
        matches!(self.robots[r].mission, Mission::Idle)
    }

    // Moves robot i one step toward (tx, ty); reports whether it now stands there.
    fn move_robot(&mut self, i: usize, tx: i64, ty: i64) -> bool {
        if self.robots[i].x == tx && self.robots[i].y == ty {
            return true;
        }
        if self.robots[i].x != tx {
            if self.robots[i].x < tx {
                self.robots[i].x += 1;
            } else {
                self.robots[i].x -= 1;
            }
        } else if self.robots[i].y < ty {
            self.robots[i].y += 1;
        } else {
            self.robots[i].y -= 1;
        }
        self.robots[i].battery -= 1;
        self.steps += 1;
        self.robots[i].x == tx && self.robots[i].y == ty
    }

    fn assign_pending(&mut self) {
        while self.pending_head < self.pending.len() {
            let oidx = self.pending[self.pending_head];
            let mut best: i64 = -1;
            let mut bestd: i64 = 1_000_000;
            for r in 0..ROBOTS {
                if self.robot_idle(r) {
                    let d = dist(
                        self.robots[r].x,
                        self.robots[r].y,
                        self.orders[oidx].px,
                        self.orders[oidx].py,
                    );
                    if d < bestd {
                        bestd = d;
                        best = r as i64;
                    }
                }
            }
            if best < 0 {
                return;
            }
            self.robots[best as usize].mission = Mission::ToPickup(oidx);
            self.pending_head += 1;
        }
    }

    fn step_robot(&mut self, i: usize) {
        let m = self.robots[i].mission;
        match m {
            Mission::Idle => {
                if self.robots[i].battery < BATTERY_LOW {
                    self.robots[i].mission = Mission::Charging;
                }
            }
            Mission::ToPickup(o) => {
                let tx = self.orders[o].px;
                let ty = self.orders[o].py;
                if self.move_robot(i, tx, ty) {
                    self.robots[i].mission = Mission::ToDropoff(o);
                }
            }
            Mission::ToDropoff(o) => {
                let tx = self.orders[o].dx;
                let ty = self.orders[o].dy;
                if self.move_robot(i, tx, ty) {
                    self.robots[i].mission = Mission::Idle;
                    self.robots[i].delivered += 1;
                    self.delivered += 1;
                }
            }
            Mission::Charging => {
                let sx = if self.robots[i].x < 32 { 0 } else { 63 };
                let sy = if self.robots[i].y < 32 { 0 } else { 63 };
                if self.move_robot(i, sx, sy) {
                    self.robots[i].battery += CHARGE_RATE;
                    if self.robots[i].battery >= BATTERY_CAP {
                        self.robots[i].battery = BATTERY_CAP;
                        self.robots[i].mission = Mission::Idle;
                    }
                }
            }
        }
    }

    fn tick(&mut self, t: i64) {
        if t % ORDER_EVERY == 0 && self.pending.len() - self.pending_head < PENDING_CAP {
            let px = self.rng.below(W);
            let py = self.rng.below(W);
            let dx = self.rng.below(W);
            let dy = self.rng.below(W);
            self.orders.push(Order { px, py, dx, dy });
            self.pending.push(self.orders.len() - 1);
        }
        self.assign_pending();
        for i in 0..ROBOTS {
            self.step_robot(i);
        }
    }
}

fn main() {
    let mut rng = Rng { state: 42 };
    let mut robots = Vec::new();
    for _ in 0..ROBOTS {
        let x = rng.below(W);
        let y = rng.below(W);
        robots.push(Robot {
            x,
            y,
            battery: BATTERY_CAP,
            delivered: 0,
            mission: Mission::Idle,
        });
    }
    let mut sim = Sim {
        robots,
        orders: Vec::new(),
        pending: Vec::new(),
        pending_head: 0,
        rng,
        delivered: 0,
        steps: 0,
    };

    for t in 0..TICKS {
        sim.tick(t);
    }

    let mut pos_hash: i64 = 0;
    let mut battery_sum: i64 = 0;
    for i in 0..ROBOTS {
        pos_hash = pos_hash
            .wrapping_mul(31)
            .wrapping_add(sim.robots[i].x * 64 + sim.robots[i].y);
        battery_sum += sim.robots[i].battery;
    }
    println!("created {}", sim.orders.len());
    println!("delivered {}", sim.delivered);
    println!("steps {}", sim.steps);
    println!("pos {}", pos_hash);
    println!("battery {}", battery_sum);
}
